Reload reducing privacy protections does not work for navigational protections after redirection

charliewolfe opened the pull request:
#### fab481e2f93023a2ade55a2e513b7f4d50305277
<pre>
Reload reducing privacy protections does not work for navigational protections after redirection
<a href="https://bugs.webkit.org/show_bug.cgi?id=258816">https://bugs.webkit.org/show_bug.cgi?id=258816</a>
rdar://111656873

Reviewed by Alex Christensen.

Some navigational privacy protections, like link decoration filtering and referrer/query string hiding,
rely on the website policy of the page that was navigated away from when determining if these
protections are enabled. Since redirects may occur between a navigation, the originating advanced
privacy protections policy should be inherited between pages unless the navigation was user initiated.

* Source/WebKit/Shared/NavigationActionData.h:
* Source/WebKit/Shared/NavigationActionData.serialization.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::decidePolicyForNavigationAction):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createWindow):
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::dispatchDecidePolicyForNavigationAction):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::didSameDocumentNavigationForFrameViaJSHistoryAPI):
(WebKit::WebLocalFrameLoaderClient::dispatchDecidePolicyForNewWindowAction):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::policySourceDocumentLoader):
(WebKit::WebFrame::advancedPrivacyProtections const):
(WebKit::WebFrame::originatorAdvancedPrivacyProtections const):
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
* Tools/TestWebKitAPI/Tests/WebKit/AdvancedPrivacyProtections.mm:
(TestWebKitAPI::webViewAfterCrossSiteNavigationWithReducedPrivacy):
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/265774@main">https://commits.webkit.org/265774@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4dd00dce5b5311ed2c1f6ca2924dda6d2654b06b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11775 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11974 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12343 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13419 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11214 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11794 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14363 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11955 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14077 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11939 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12773 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9991 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13840 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10065 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10690 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17820 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11142 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10845 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14015 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11251 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9293 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10425 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2855 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14708 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11107 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -1144,25 +1144,42 @@ std::optional<NavigatingToAppBoundDomain> WebFrame::isTopFrameNavigatingToAppBou
+
+OptionSet<WebCore::AdvancedPrivacyProtections> WebFrame::advancedPrivacyProtections() const
+{
+    auto* loader = policySourceDocumentLoader(this->coreLocalFrame());
```
I don't think there's a good reason to pass a LocalFrame* in as a parameter.

sysrqb commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -1144,25 +1144,43 @@ std::optional<NavigatingToAppBoundDomain> WebFrame::isTopFrameNavigatingToAppBou
-        return { };
+        return nullptr;
 
     auto* policySourceDocumentLoader = topDocumentLoader;
```
I know this was pre-existing, but why don't we just name this `policySourceDocumentLoader` above?

charliewolfe replied:
That does seem better, fixed.

sysrqb commented on Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
```diff
@@ -142,6 +142,7 @@ void WebFrameLoaderClient::dispatchDecidePolicyForNavigationAction(const Navigat
         coreFrame ? coreFrame->loader().effectiveSandboxFlags() : SandboxFlags(),
         navigationAction.privateClickMeasurement(),
         requestingFrame ? requestingFrame->advancedPrivacyProtections() : OptionSet<AdvancedPrivacyProtections> { },
+        requestingFrame ? requestingFrame->originatorAdvancedPrivacyProtections() : OptionSet<AdvancedPrivacyProtections> { },
```
Do (or will) we need this information for anything else in the future? Could you just do something like, instead:
```
requestingFrame ? (redirectResponse.isNull() ? requestingFrame->advancedPrivacyProtections() : requestingFrame->originatorAdvancedPrivacyProtections()) : OptionSet<AdvancedPrivacyProtections> { },
```

This seems like it would avoid some of the complexity.

sysrqb replied:
I realize this wasn't exactly clear. I mean instead of adding the new entry, just set the `advancedPrivacyProtections` entry with the correct value (if this is a correct assumption).

charliewolfe replied:
I don't think that will work, since we want to cover both client and server side redirection. Checking the response to see if there was a redirect will only cover server side redirection.

The tests which use `window.location = x` cover this.

sysrqb replied:
Okay, that makes sense. There's probably another way to do this from WebCore based on FrameLoader::changeLocation, but your patch looks fine.

charliewolfe committed:
Reload reducing privacy protections does not work for navigational protections after redirection
https://bugs.webkit.org/show_bug.cgi?id=258816
rdar://111656873

Reviewed by Alex Christensen.

Some navigational privacy protections, like link decoration filtering and referrer/query string hiding,
rely on the website policy of the page that was navigated away from when determining if these
protections are enabled. Since redirects may occur between a navigation, the originating advanced
privacy protections policy should be inherited between pages unless the navigation was user initiated.

* Source/WebKit/Shared/NavigationActionData.h:
* Source/WebKit/Shared/NavigationActionData.serialization.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::decidePolicyForNavigationAction):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createWindow):
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::dispatchDecidePolicyForNavigationAction):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::didSameDocumentNavigationForFrameViaJSHistoryAPI):
(WebKit::WebLocalFrameLoaderClient::dispatchDecidePolicyForNewWindowAction):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::policySourceDocumentLoader):
(WebKit::WebFrame::advancedPrivacyProtections const):
(WebKit::WebFrame::originatorAdvancedPrivacyProtections const):
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
* Tools/TestWebKitAPI/Tests/WebKit/AdvancedPrivacyProtections.mm:
(TestWebKitAPI::webViewAfterCrossSiteNavigationWithReducedPrivacy):
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/265774@main
